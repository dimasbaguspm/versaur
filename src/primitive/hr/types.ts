import type { HTMLAttributes } from 'react'

/**
 * HrProps defines the props for the Hr component
 * @property hasMargin - Whether to include bottom margin (mb-4)
 */
export interface HrProps extends HTMLAttributes<HTMLHRElement> {
  /**
   * Whether to include bottom margin
   * - When true: Adds mb-4 for spacing
   * - When false: No margin applied (default)
   */
  hasMargin?: boolean
}
