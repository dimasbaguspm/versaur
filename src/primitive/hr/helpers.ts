import { cva } from '@/utils/variants'

/**
 * Hr component variants using Versaur color system
 * Creates a horizontal line with border color and optional margin
 */
export const hrVariants = cva('h-px w-full bg-border border-0', {
  variants: {
    hasMargin: {
      true: 'mb-4',
      false: '',
    },
  },
  defaultVariants: {
    hasMargin: false,
  },
})
