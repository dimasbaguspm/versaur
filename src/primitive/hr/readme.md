# Hr Component

A simple horizontal line separator component for creating visual divisions in content.

## Usage

```tsx
import { Hr } from '@versaur/primitive'

// Basic usage
<Hr />

// With bottom margin
<Hr hasMargin />

// With custom styling
<Hr className="my-8 opacity-50" />
```

## Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `hasMargin` | `boolean` | `false` | Whether to include bottom margin (mb-4) |
| `className` | `string` | - | Additional CSS classes |

## Features

- 1px height horizontal line
- Uses `bg-border` color from the design system
- Optional bottom margin (`mb-4`)
- Fully accessible as standard HTML `<hr>` element
- Supports all standard HTMLHRElement props
