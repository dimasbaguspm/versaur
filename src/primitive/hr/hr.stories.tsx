import type { Meta, StoryObj } from '@storybook/react'
import { Hr } from './hr'

const meta: Meta<typeof Hr> = {
  title: 'Primitive/Hr',
  component: Hr,
  parameters: {
    layout: 'padded',
  },
  argTypes: {
    hasMargin: {
      control: 'boolean',
      description: 'Whether to include bottom margin (mb-4)',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

/**
 * Default horizontal line separator
 */
export const Default: Story = {}

/**
 * Horizontal line with bottom margin for spacing
 */
export const WithMargin: Story = {
  args: {
    hasMargin: true,
  },
}

/**
 * Example showing Hr usage in content sections
 */
export const InContent: Story = {
  render: () => (
    <div className='space-y-4'>
      <div>
        <h3 className='text-lg font-semibold'>Section 1</h3>
        <p className='text-sm text-foreground-light'>
          This is some content in the first section.
        </p>
      </div>
      <Hr hasMargin />
      <div>
        <h3 className='text-lg font-semibold'>Section 2</h3>
        <p className='text-sm text-foreground-light'>
          This is some content in the second section.
        </p>
      </div>
      <Hr hasMargin />
      <div>
        <h3 className='text-lg font-semibold'>Section 3</h3>
        <p className='text-sm text-foreground-light'>
          This is some content in the third section.
        </p>
      </div>
    </div>
  ),
}
