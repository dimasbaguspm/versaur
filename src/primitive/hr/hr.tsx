import { forwardRef } from 'react'
import { cn } from '@/utils/cn'
import { hrVariants } from './helpers'
import type { HrProps } from './types'

/**
 * Hr component for creating a horizontal line separator with optional margin
 *
 * Follows the Regular Pattern - a simple styled wrapper that aligns with standard HTML behavior.
 * Creates a 1px height horizontal line with border color and optional bottom margin.
 *
 * @example
 * ```tsx
 * // Basic usage
 * <Hr />
 *
 * // With bottom margin
 * <Hr hasMargin />
 *
 * // With custom className
 * <Hr className="my-8" />
 * ```
 */
export const Hr = forwardRef<HTMLHRElement, HrProps>(
  ({ hasMargin = false, className, ...props }, ref) => {
    return (
      <hr
        ref={ref}
        className={cn(hrVariants({ hasMargin }), className)}
        {...props}
      />
    )
  }
)
