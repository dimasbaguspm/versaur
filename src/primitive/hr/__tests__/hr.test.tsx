import { render } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../hr.stories'

const { Default, WithMargin } = composeStories(stories)

describe('Hr', () => {
  it('renders correctly with default props', () => {
    const { container } = render(<Default />)
    const hr = container.querySelector('hr')

    expect(hr).toBeInTheDocument()
    expect(hr).toHaveClass('h-px', 'w-full', 'bg-border', 'border-0')
    expect(hr).not.toHaveClass('mb-4')
  })

  it('renders with margin when hasMargin is true', () => {
    const { container } = render(<WithMargin />)
    const hr = container.querySelector('hr')

    expect(hr).toBeInTheDocument()
    expect(hr).toHaveClass('h-px', 'w-full', 'bg-border', 'border-0', 'mb-4')
  })

  it('applies custom className correctly', () => {
    const { container } = render(<Default className='my-custom-class' />)
    const hr = container.querySelector('hr')

    expect(hr).toHaveClass('my-custom-class')
  })

  it('forwards ref correctly', () => {
    const ref = { current: null }
    render(<Default ref={ref} />)

    expect(ref.current).toBeInstanceOf(HTMLHRElement)
  })

  it('matches snapshot', () => {
    expect(render(<Default />).asFragment()).toMatchSnapshot()
  })
})
