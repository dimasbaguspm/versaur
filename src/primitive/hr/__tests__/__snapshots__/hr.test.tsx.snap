// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Hr > matches snapshot 1`] = `
<DocumentFragment>
  <hr
    class="h-px w-full bg-border border-0"
  />
</DocumentFragment>
`;
